### PR TITLE
Feature/result mapping

### PR DESCRIPTION
**Pull request by Insomniak47**
+semver:feature

Added result mapping extensions and wrote some of the most boring and hard to name tests ever. Cleaned up some of the other tests where I noticed coverage was missing and expanded OkOr with two methods that collapse Result<Result<T>> so that we don't have to collapse manually. Probably the most boring 97 LOC I've ever written but they let me do this: 

image

image

### CODE REVIEW

**Comment by Insomniak47** [comment]
closes #39

File: wimm.Secundatives.UnitTests/Extensions/Expect_Test.cs [hunk 1]
```diff
@@ -82,6 +82,16 @@ public async Task ExpectTaskMessage_NoValue_ThrowsWithMessage()
             Assert.Equal(_message, ex.Message);
         }
 
+        [Fact]
+        public async Task ExpectTaskMessage_Value_ThrowsWithMessage()
```
Comment (slagroom) [inline wimm.Secundatives.UnitTests/Extensions/Expect_Test.cs]: This was very disorienting... shouldn't it be `ExpectTaskMessage_Value_ReturnsValue` and 

```
Assert.Equal(10, await underTest.Expect(_message);
```

?

File: wimm.Secundatives.UnitTests/Extensions/OkOr_Test.cs [hunk 2]
```diff
@@ -89,6 +89,24 @@ public async Task OkOrErrorTaskValue_None_ReturnsErrorValue()
             Assert.Equal(TestingEnum.BadThingsHappened, result.Error);
         }
 
+        [Fact] 
+        public async Task OkOrErrorTaskValue_ResultMemberExists_ReturnsInternalResult()
```
Comment (slagroom) [inline wimm.Secundatives.UnitTests/Extensions/OkOr_Test.cs]: Are you missing a test for value/none and func/some ?

File: wimm.Secundatives.UnitTests/Extensions/ResultMapping_Test.cs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        }
+
+        [Fact]
+        public void MapResult_IsErrorAndFuncReturnsValue_ReturnsError()
+        {
+            var underTest = new Result<string, TestError>(TestError.Sadness);
+
+            var res = underTest.Map(x => "deet");
+
+            Assert.True(res.IsError);
```
Comment (slagroom) [inline wimm.Secundatives.UnitTests/Extensions/ResultMapping_Test.cs]: The is error test is redundant since accessing Error on a non-error will throw

File: wimm.Secundatives.UnitTests/Extensions/ResultMapping_Test.cs [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+            Assert.Equal(TestError.Sadness, res.Error);
+
+        }
+
+        [Fact]
+        public async Task MapResult_IsSuccessAndFuncReturnsTask_ReturnsResultOfFunc()
+        {
+            var underTest = ConstructWith("doot");
+            var res = await underTest.Map(async x => await Task.FromResult("skeleton"));
```
Comment (slagroom) [inline wimm.Secundatives.UnitTests/Extensions/ResultMapping_Test.cs]: not that it matters, but is there any reason in the tests to do 
```
var res = await underTest.Map(async x => await Task.FromResult("skeleton"));
```
over just
```
var res = await underTest.Map(x => Task.FromResult("skeleton"));
```
?

File: wimm.Secundatives.UnitTests/Extensions/ResultMapping_Test.cs [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        public async Task MapTaskResult_IsErrorAndFuncReturnsTask_ReturnsError()
+        {
+            var underTest = Task.FromResult(new Result<string, TestError>(TestError.Sadness));
+            var res = await underTest.Map(async x => await Task.FromResult("skeleton"));
+
+            Assert.True(res.IsError);
+            Assert.Equal(TestError.Sadness, res.Error);
+        }
+
+        //
```
Comment (slagroom) [inline wimm.Secundatives.UnitTests/Extensions/ResultMapping_Test.cs]: //

File: wimm.Secundatives.UnitTests/Extensions/ResultMapping_Test.cs [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+        [Fact]
+        public async Task MapResult_IsErrorAndFuncReturnsTaskResult_ReturnsError()
+        {
+            var underTest = new Result<string, TestError>(TestError.Sadness);
+            var res = await underTest.Map(async x => await Task.FromResult(ConstructWith("skeleton")));
+
+            Assert.True(res.IsError);
+            Assert.Equal(TestError.Sadness, res.Error);
+        }
+        //
```
Comment (slagroom) [inline wimm.Secundatives.UnitTests/Extensions/ResultMapping_Test.cs]: //

File: wimm.Secundatives.UnitTests/Extensions/ResultMapping_Test.cs (hunk 3 above)
Comment (Insomniak47) [reply wimm.Secundatives.UnitTests/Extensions/ResultMapping_Test.cs]: Was more about being clear in the assertions than anything. Want to fail the assert not just throw

File: wimm.Secundatives.UnitTests/Extensions/OkOr_Test.cs (hunk 2 above)
Comment (Insomniak47) [reply wimm.Secundatives.UnitTests/Extensions/OkOr_Test.cs]: I am

Comment (Insomniak47) [reply wimm.Secundatives.UnitTests/Extensions/OkOr_Test.cs]: wf

File: wimm.Secundatives.UnitTests/Extensions/ResultMapping_Test.cs (hunk 4 above)
Comment (Insomniak47) [reply wimm.Secundatives.UnitTests/Extensions/ResultMapping_Test.cs]: I was super tired

File: wimm.Secundatives.UnitTests/Extensions/Expect_Test.cs (hunk 1 above)
Comment (Insomniak47) [reply wimm.Secundatives.UnitTests/Extensions/Expect_Test.cs]: Yes. Brain dead while finishing this up. good catch

File: wimm.Secundatives.UnitTests/Extensions/ResultMapping_Test.cs (hunk 5 above)
Comment (Insomniak47) [reply wimm.Secundatives.UnitTests/Extensions/ResultMapping_Test.cs]: wf

File: wimm.Secundatives.UnitTests/Extensions/ResultMapping_Test.cs (hunk 6 above)
Comment (Insomniak47) [reply wimm.Secundatives.UnitTests/Extensions/ResultMapping_Test.cs]: wf

File: wimm.Secundatives.UnitTests/Extensions/Expect_Test.cs [hunk 7]
```diff
@@ -83,7 +80,7 @@ public async Task ExpectTaskMessage_NoValue_ThrowsWithMessage()
         }
 
         [Fact]
-        public async Task ExpectTaskMessage_Value_ThrowsWithMessage()
+        public async Task ExpectTaskMessage_Value_ReturnsValue()
         {
             var underTest = Task.FromResult(new Maybe<int>(10));
             var ex = await underTest.Expect(_message);
```
Comment (slagroom) [inline wimm.Secundatives.UnitTests/Extensions/Expect_Test.cs]: Calling this `ex` is a bit weird too
